pkg/controller/installation/products/threescale/reconciler.go

```diff
@@ -332,6 +299,146 @@ func (r *Reconciler) reconcileComponents(ctx context.Context, serverClient pkgcl
 	return v1alpha1.PhaseInProgress, nil
 }
 
+func (r *Reconciler) reconcileBlobStorage(ctx context.Context, serverClient pkgclient.Client) (v1alpha1.StatusPhase, error) {
+	logrus.Info("Reconciling blob storage")
+
+	// setup blob storage cr for the cloud resource operator
+	blobStorageName := fmt.Sprintf("3scale-s3-%s", r.installation.Name)
+	blobStorage := &crov1.BlobStorage{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      blobStorageName,
+			Namespace: r.installation.Namespace,
+		},
+	}
+
+	logrus.Info("Creating blob storage")
+	_, err := controllerutil.CreateOrUpdate(ctx, serverClient, blobStorage, func(existing runtime.Object) error {
+		c := existing.(*crov1.BlobStorage)
+		c.Spec.Type = r.installation.Spec.Type
+		c.Spec.Tier = "production"
+		c.Spec.SecretRef = &crov1.SecretRef{
+			Name:      fmt.Sprintf("3scale-s3-%s", r.installation.Name),
+			Namespace: r.Config.GetNamespace(),
+		}
+		return nil
+	})
+	if err != nil {
+		return v1alpha1.PhaseFailed, errors.Wrap(err, "failed to reconcile blob storage request")
+	}
+
+	// wait for the blob storage cr to reconcile
+	if blobStorage.Status.Phase != crov1.PhaseComplete {
+		return v1alpha1.PhaseAwaitingComponents, nil
+	}
+
+	return v1alpha1.PhaseCompleted, nil
+}
+
+func (r *Reconciler) getFileStorageSpec(ctx context.Context, serverClient pkgclient.Client) (*threescalev1.SystemFileStorageSpec, error) {
+	// if cro is being used for blob storage
+	if r.installation.Spec.UseExternalResources {
+		// create blob storage cr
+		blobStorage := &crov1.BlobStorage{}
+		err := serverClient.Get(ctx, pkgclient.ObjectKey{Name: fmt.Sprintf("3scale-s3-%s", r.installation.Name), Namespace: r.installation.Namespace}, blobStorage)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to get blob storage custom resource")
+		}
+
+		// get blob storage connection secret
+		blobStorageSec := &v1.Secret{}
+		err = serverClient.Get(ctx, pkgclient.ObjectKey{Name: blobStorage.Name, Namespace: r.Config.GetNamespace()}, blobStorageSec)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to get blob storage connection secret")
+		}
+
+		// create s3 credentials secret
+		credSec := &v1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      s3CredentialsSecretName,
+				Namespace: r.Config.GetNamespace(),
+			},
+			Data: map[string][]byte{},
+		}
+		err = serverClient.Get(ctx, pkgclient.ObjectKey{Name: credSec.Name, Namespace: credSec.Namespace}, credSec)
+		if err != nil && k8serr.IsNotFound(err) {
+			credSec.Data["AWS_ACCESS_KEY_ID"] = blobStorageSec.Data["credentialKeyID"]
+			credSec.Data["AWS_SECRET_ACCESS_KEY"] = blobStorageSec.Data["credentialSecretKey"]
+
+			err = serverClient.Create(ctx, credSec)
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to create blob storage aws credentials secret")
+			}
+		}
+
+		// return the file storage spec
```

Comment:
From here on, we could do

- read operator sec
- createorupdate namespace sec using data from operator sec

that will flatten this entire section

i'd also split these into two funcs and move the external resource conditional up a level